Add MQTT-SN library

BeckmaR opened the pull request:
This library implements MQTT-SN version 1.2. It is transport agnostic;
users will have to implement a transport themselves. This make sense
since MQTT-SN does not require any kind of transport or protocol, it
just has to be frame-based.

The library uses a k_heap for dynamic memory allocation. This allows to
make use of the larger messages introduced in v1.2, without requiring
enormous amounts of memory to be pre-allocated.

This library has not been tested with an actual gateway. There are
however unit tests that make sure that package encoding and decoding
work correctly, as well as some basic connect and publish logic.

Signed-off-by: Rene Beckmann <rene.bckmnn@gmail.com>


---------
I have worked on this on and off for the last months. I would love to get some feedback on this - is it generally interesting, should I do something fundamentally different, does it make sense?

carlescufi commented:
Thanks for the PR!
I am not an MQTT expert, but is there a reason why you introduce a new MQTT implementation instead of extending the existing one?

rlubos commented on include/net/mqtt_sn.h:
```diff
@@ -0,0 +1,197 @@
+ */
+struct mqtt_sn_callbacks {
+	/** Will be called by the library when it wants to send a message. */
+	int (*msg_send)(struct mqtt_sn_msg *msg, const void *recipient);
```
I see that the approach here regarding communication (socket) handling is slightly different than in the core MQTT lib we have. Here, we expect the application to handle the communication part, while the standard MQTT lib handles sockets on its own, with an abstracted transport layer. I wonder if it made sense to follow that approach here, to give users unanimous feel of both libraries.

BeckmaR replied:
This was a design decision I made because MQTT-SN does not require any specific kind of transport, unlike MQTT which does require a reliable transport like TCP / TLS. You can use UDP for MQTT-SN, but you could also use ZigBee, UART, GATT, LoRaWAN or potentially even SMS.

It might make sense to take inspiration from the original MQTT library to have some kind of transport-related struct of function pointers, and provide a default one for UDP. That would make the approach similar, even if the details will differ.

BeckmaR replied:
@rlubos The more I go into the details _how_ the MQTT library handles the transport, the less I like it, actually. The user is expected to call `mqtt_input` when there is new data. To know if that is the case, the user of course has to poll the socket. This breaks the separation of concerns IMO (see `zephyr/samples/net/cloud/google_iot_mqtt/src/protocol.c:215`, `wait_for_input`).

rlubos replied:
Well, there was an argument that it should be possible to run MQTT along with other protocols on a single thread, therefore `mqtt_input()` is non-blocking and relies on the application to actually monitor the activity on a socket (so that it can add whatever sockets it wants to the `fds`). But perhaps it wouldn't be a bad idea to have some kind of `mqtt_wait()` function.

rlubos commented on include/net/mqtt_sn.h:
```diff
@@ -0,0 +1,197 @@
+	/** Will be called by the library when it wants to broadcast a message. */
+	int (*msg_broadcast)(struct mqtt_sn_msg *msg, uint8_t radius);
+	/** Will be called by the library when it receives a GWINFO message. */
+	void (*gwinfo)(uint8_t gwid, struct mqtt_sn_data *gw_add);
+	/** Will be called by the library when it receives a PUBLISH message. */
+	void (*publish)(struct mqtt_sn_data *data, enum mqtt_sn_topic_type type, uint16_t tid);
```
Again, this is slightly different from our MQTT library, where we only register a single callback, which can handle different event types. I think we could do the same here, to remain consistent.

BeckmaR replied:
That might make sense, even though this library exposes fewer events than the original MQTT library. For example, I handle stuff like ACKs internally, where the MQTT library leaves that as an exercise to the user.

rlubos commented on include/net/mqtt_sn.h:
```diff
@@ -0,0 +1,197 @@
+
+	int state;
+
+	struct k_work_delayable wrk;
```
Could use a more meaningful name here.

rlubos commented on include/net/mqtt_sn.h:
```diff
@@ -0,0 +1,197 @@
+/**
+ * Quality of Service. QoS 0-2 work the same as basic MQTT, QoS -1 is an MQTT-SN addition.
+ */
+enum __attribute__((packed)) mqtt_sn_qos {
```
Why do they need to be packed?

jukkar replied:
And if it really needs to be packed, there is the `__packed` define that can be used here.

BeckmaR replied:
It does make the parameter-union in the encoder and decoder parts much smaller. In an earlier version I had to store those for re-sends. Now, they are still used in the `mqtt_sn_publish` and `mqtt_sn_topic` structs. I thought using 3 bytes instead of 12 might be useful (lots of assumptions here).

BeckmaR replied:
removed these

rlubos commented on subsys/net/lib/mqtt_sn/Kconfig:
```diff
@@ -0,0 +1,52 @@
+# Socket MQTT Library for Zephyr
+
+# Copyright (c) 2018 Nordic Semiconductor ASA
```
Please update copyright (here and some other files)

BeckmaR replied:
Done.

rlubos reviewed (commented):
Thank you for this contribution, I have to say it looks quite impressive. I've left a few inital high-level comments.

@carlescufi I see that the overall design of the library is different from the core MQTT library, but I think that it's somewhat justified with MQTT-SN:
1. The library encapsulates retransmission logic within, as MQTT-SN is built upon non-reliable UDP transport, I think it's the right choice to hande this within the library.
2. I was a bit reluctant initially about the heap usage in the library, but after a closer look I think it makes sense in this case - especially that it has a dedicated heap. First, we need memory to store the messages for reatransmission, so we can't just rely on stack here. Second, with UDP we can't send packets partially as we do in regular MQTT, we need to compose header/payload into a single linear buffer before transmission, due to the nature of UDP sockets.
3. And generally, while the general idea behind MQTT-SN is similar to MQTT, we're not interfacing MQTT Broker directly here, but rather an intermediate MQTT Gateway - the communication scheme between client/gateway is a bit different than between client/broker. I'm not really sure if extending the existing implementation to match SN requirements would be a viable choice here.

I would consider though to relieve the application from handling the sockets on it's own as we do in the regular MQTT lib. It's up to discussion though, if there are arguments not to do so I can reconsider.

BeckmaR commented on subsys/net/lib/mqtt_sn/mqtt_sn.c:
```diff
@@ -0,0 +1,1115 @@
+{
+	struct mqtt_sn_publish *pub;
+
+	// Destroy all pubs referencing this topic
```
Done, hope I got all.

BeckmaR commented:
@carlescufi In the end, MQTT-SN and MQTT are different enough in some aspects that I thought it would really be more confusing to have them side-by-side in the same library than just build a completely separate one. Also there are some aspects about the original library that I don't like, e.g. that it does not handle ACKs and PUBREC etc itself.

rlubos commented on subsys/net/lib/mqtt_sn/mqtt_sn.c:
```diff
@@ -0,0 +1,1150 @@
+	mqtt_sn_msg_unref(msg);
+}
+
+static void handle_pingresp(struct mqtt_sn_client *client)
```
Could we drop the empty functions? I don't see a reason to introduce them, if particular replies don't require any processing.

rlubos commented on include/zephyr/net/mqtt_sn.h:
```diff
@@ -0,0 +1,290 @@
+			struct mqtt_sn_transport *transport, mqtt_sn_evt_cb_t evt_cb);
+
+#ifdef CONFIG_MQTT_SN_TRANSPORT_UDP
+int mqtt_sn_client_init_udp(struct mqtt_sn_client *client, const struct mqtt_sn_data *client_id,
```
Is this function actually implemented anywhere?

BeckmaR replied:
no, removed this

rlubos commented on subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:
```diff
@@ -0,0 +1,106 @@
+}
+
+static int tp_udp_msg_send(struct mqtt_sn_client *client, struct mqtt_sn_msg *msg,
+	     const void *recipient)
```
The `recepient` field does not seem to be used, what's its purpose?

BeckmaR replied:
removed

rlubos commented on subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c:
```diff
@@ -0,0 +1,106 @@
+	struct mqtt_sn_transport_udp *udp = UDP_CONTAINER(transport);
+	int err;
+
+	udp->sock = zsock_socket(PF_INET, SOCK_DGRAM, 0);
```
There should also be some function that closes the socket. In the regular `MQTT` implementation the approach is a bit different, the socket is created on "connect" and closed on "disconnect" calls, but it's a bit different due to TCP (the socket is no longer usable after TCP disconnect). Perhaps we need some kind of `deinit()` function here?

rlubos reviewed (commented):
The library/test looks pretty good, I assume the sample part is still in early WIP?

BeckmaR commented:
Hey @rlubos, thank you for the review. Due to multiple reasons I didn't get around to working on this in the meantime. I have now started to work on it again (fingers crossed) and changed some things that I did not like before:
- Removed the useless `struct mqtt_sn_msg` and everything related to it - the encoder and decoder now just use a `net_buf_simple`
- Removed the terrible heap, I aligned with libraries like lwm2m and mqtt and simply introduced n times x bytes of memory (globally) for publish data and topics
- Removed the gateway void pointer. I decided not to support broadcasting and SEARCHGW currently, users will just have to provide e.g. a gateway IP and port to the UDP transport, or take care of that somewhere else when using e.g. Bluetooth

I will now take your other comments into account and e.g. remove the empty functions. Next step would be to create a sample application and test this against an actual gateway.

BeckmaR commented on include/zephyr/net/mqtt_sn.h:
```diff
@@ -0,0 +1,290 @@
+
+	int state;
+
+	struct k_work_delayable wrk;
```
renamed to `process_work`

rlubos reviewed (commented):
This looks quite neat, thank you for the work you put into this library. I have a few final requests:
* Please split the library/sample/tests into separate commits, each commit following the [commit guidelines](https://docs.zephyrproject.org/3.2.0/contribute/guidelines.html#commit-guidelines) (title/body/signoff)
* There are quite a few style issues, please check https://github.com/zephyrproject-rtos/zephyr/actions/runs/3329890091/jobs/5507679599 for details.
* Looks that some minor fixes are needed in the documentation as well: https://github.com/zephyrproject-rtos/zephyr/actions/runs/3329890086/jobs/5507679754#step:7:51
* Please add `sample.yaml` to the sample, so that it'll be built by the CI (here's an example: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/mqtt_publisher/sample.yaml)
* And finally, please add an entry for the library in the [CODEOWNERS](https://github.com/zephyrproject-rtos/zephyr/blob/main/CODEOWNERS) and [MAINTAINERS](https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml) file.

BeckmaR commented on samples/net/mqtt_sn_publisher/src/common.h:
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2022 grandcentrix GmbH
```
That is intentional since I did some parts on company time. Thank you for watching out!

BeckmaR commented on tests/net/lib/mqtt_sn_client/prj.conf:
```diff
@@ -0,0 +1,23 @@
+CONFIG_NET_LOG=y
+CONFIG_MQTT_SN_LOG_LEVEL_DBG=y
+
+CONFIG_COMPILER_OPT="-Werror -Wno-unused-function -Wno-unused-variable"
```
Ah, it was a crudge to compile the library with these options in the beginning - leftover.

BeckmaR commented:
@rlubos Still polishing some stuff, some tests fail due to minor errors. If everything builds I will split it up into smaller commits.

I can't figure out why the documentation build fails, it complains it can't find LaTeX - I didn't touch anything in that area.

BeckmaR committed:
net: mqtt-sn: Add MQTT-SN library

This commit adds an implementation of MQTT-SN v1.2.
The specification is available on oasis-open.org:
https://www.oasis-open.org/committees/download.php/66091/MQTT-SN_spec_v1.2.pdf

The following things are missing in this implementation:
- Pre-defined topic IDs
- QoS -1 - it's most useful with predefined topics
- Gateway discovery using ADVERTISE, SEARCHGW and GWINFO messages.
- Setting the will topic and message after the initial connect
- Forwarder Encapsulation

Signed-off-by: René Beckmann <rene.beckmann@grandcentrix.net>

BeckmaR committed:
samples: net: mqtt-sn: Add MQTT-SN publisher sample

This sample demonstrates the usage of the MQTT-SN library over UDP.
It connects to a gateway, publishes to the topic "/uptime" and also
subscribes to "/number".

Signed-off-by: René Beckmann <rene.beckmann@grandcentrix.net>

BeckmaR committed:
tests: net: mqtt-sn: Add encoding/decoding test

Validate that the encoding and decoding part of the MQTT-SN library
work correctly using prepared byte arrays of messages.

Signed-off-by: René Beckmann <rene.beckmann@grandcentrix.net>

BeckmaR committed:
tests: net: mqtt-sn: Add test for MQTT-SN client

Validate that basic lifecycling and usage of the MQTT-SN client
work correctly by providing a mocked MQTT-SN transport. Connection
setup and state changes are tested.

Signed-off-by: René Beckmann <rene.beckmann@grandcentrix.net>

BeckmaR committed:
doc: net: mqtt-sn: Add documentation for MQTT-SN library

Add a page explaining the usage of the MQTT-SN library, the client
and the UDP transport, and add it to the `protocols` index page
next to MQTT.

Signed-off-by: René Beckmann <rene.beckmann@grandcentrix.net>

BeckmaR committed:
CODEOWNERS, MAINTAINERS: Add entries for MQTT-SN

This adds the files related to MQTT-SN, with an additional exlude
to separate MQTT and MQTT-SN.

Signed-off-by: René Beckmann <rene.beckmann@grandcentrix.net>

BeckmaR commented:
@rlubos CI is happy, as mentioned above I have no idea about what you expect for the CODEOWNERS and MAINTAINERS files.

rlubos commented:
> @rlubos CI is happy, as mentioned above I have no idea about what you expect for the CODEOWNERS and MAINTAINERS files.

The changes in those files are just what I expected, thanks. @carlescufi Is the GH warning from the CODEOWNERs file something we should be concerned about? I see that there are already entries like the one added, where not all users have write access to the repo.

rlubos reviewed (dismissed):
LGTM, thank you for the PR, I have no more concerns.

stephanosio commented:
> The changes in those files are just what I expected, thanks. @carlescufi Is the GH warning from the CODEOWNERs file something we should be concerned about? I see that there are already entries like the one added, where not all users have write access to the repo.

No, since the user has been added to the MAINTAINERS file (which is the authoritative one) as a collaborator, the user will be granted the collaborator access once this PR is merged and the warning will no longer apply.

jukkar commented on samples/net/mqtt_sn_publisher/Kconfig:
```diff
@@ -0,0 +1,18 @@
+
+mainmenu "MQTT-SN sample application"
+
+config MQTT_SN_GATEWAY_IP
+	string "IP of the MQTT-SN gateway"
+
+config MQTT_SN_GATEWAY_PORT
+	int "Port of the MQTT-SN gateway"
+
+config MQTT_SN_BUFFER_SIZE
```
For the sample apps, I would suggest that `SAMPLE` is prefixed to config symbols so `SAMPLE_MQTT_SN_GATEWAY_IP` etc. This makes it clear that the symbol is only available for this sample. See some other network samples for examples how to do it.

BeckmaR replied:
Good idea - done.

rlubos commented on subsys/net/lib/mqtt_sn/Kconfig:
```diff
@@ -0,0 +1,59 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config MQTT_SN_LIB
+	bool "MQTT-SN Library Support"
```
That's a good idea, we could keep the library experimental for a few releases so that we have more freedom finetuning the API if needed. @BeckmaR Here's an example on how to set a library as experimental: https://github.com/zephyrproject-rtos/zephyr/blob/f483b1bc4cfe2381624fac6337cc74d3ce2bc850/subsys/net/lib/tftp/Kconfig#L6 (note the `select EXPERIMENTAL` is also needed).

BeckmaR replied:
Done that, makes sense. Thanks for the input.

jukkar reviewed (dismissed):
LGTM
